### ISSUE TITLE
Set global Ractive variable even when AMD is detected

### ISSUE THREAD

**Issue by Ghazgkull** [ractivejs/ractive#2270]
It would be very helpful if Ractive always set the global Ractive property (e.g. window.Ractive) when it's loaded. Currently, the global property isn't set if define.amd or module.exports is detected.

The current behavior works fine for monolithic environments, where authors can make their own decision to use a module loader like RequireJS or not. But the current behavior doesn't work in the messy reality where Ractive is pulled into a mix of other 3rd party scripts.

The problem case looks like this:
1. Write a plain Javascript library that doesn't use RequireJS.
2. Load Ractive from your Javascript library by adding the appropriate `<script>` tag to the document.
3. Access Ractive in your library via window.Ractive. All is well. You now have a library that you can publish for people to include in their own sites.
4. ...until you're loaded onto a page with some other script that happens to use RequireJS. At this point, a global define.amd will become available, Ractive will see it and stop setting the window.Ractive property. Even worse, Ractive is defined as an anonymous module in RequireJS so there's literally no way to access it if it was loaded via a `<script>` tag instead of through the RequireJS API.

The problem with RequireJS which prevents accessing Ractive if it's loaded via a `<script>` tag is described on the RequireJS site:

http://requirejs.org/docs/errors.html#mismatch

> Be sure to load all scripts that call define() via the RequireJS API. Do not manually code script tags in HTML to load scripts that have define() calls in them.

Basically, RequireJS adopts an "all or nothing" model where once they're loaded on a page, they assume that all JS on the page is using them. But libraries like Ractive can make sure consumers still have a choice by continuing to set a global property in addition to the module system integrations.

**Comment by Ghazgkull** [comment]
A little extra context. [Here's a great article](http://ifandelse.com/its-not-hard-making-your-library-support-amd-and-commonjs/#amdcommonjsandplainbrowsersupport) talking about adding AMD/CommonJS support to your module, which includes a discussion at the end in the [Update section](http://ifandelse.com/its-not-hard-making-your-library-support-amd-and-commonjs/#update) about why they added back the setting of a global variable.

**Comment by JonDum** [comment]
I thoroughly disagree with this. What if you're loading a self-contained Ractive app with a different version of Ractive then what's on the global? What if you need (for whatever crazy reason) to have different defaults, decorators, components on different Ractive instances? 

You don't gain anything by forcing Ractive to add itself the global, you only lose flexibility that's up to the programmer's discretion. You can always export Ractive to the global after importing it with your module loader if that's what you want.

**Comment by Rich-Harris** [comment]
Unfortunately there's no right answer – I've experienced both situations outlined above. (To be fair, it's the situation where the global *isn't* exposed that's the most hair-pulling, in my experience.)

If we *did* force Ractive to expose a global, we would need to add back `Ractive.noConflict()` (#2066).

Ractive is built with Rollup – coincidentally there was an issue the other day on Rollup that's related to this: https://github.com/rollup/rollup/issues/269

**Comment by Rich-Harris** [comment]
Meant to add: perhaps the only solution that ultimately works for everyone is to offer different flavours of the build – an AMD-only one, and CommonJS-only one, a globals-only one, and an ES6-only one, alongside the default UMD build. Open question whether they should be added to the package by default or created via some sort of custom build generator website

**Comment by Ghazgkull** [comment]
@Rich-Harris Yeah, I'm the same person who asked for Ractive.noConflict() back. :-)

This issue is worse than the noConflict situation, though. That is something that a client can pretty easily workaround themselves, by stashing any existing version of window.Ractive before loading their own version and then restoring it after the new script loads.

Due to the RequireJS's religious stance on this issue, there's no such workaround available for this problem. What I've had to do as a workaround is to host my own version of the Ractive library which hacks around the problem (wrapping the library in a closure that just contains `var define = undefined;`).

What Ractive is doing is the standard boilerplate code for supporting AMD/CommonJS, but other libraries have evolved that pattern successfully as the link in my first comment illustrates. jQuery is a good example; they always expose a global in addition to supporting AMD/CommonJS... along with noConflict as you observe.

**Comment by martypdx** [comment]
> perhaps the only solution that ultimately works for everyone is to offer different flavours of the build

Seems like the only way to satisfy the "must have global" and "would be horrified if made global" crowds.

**Comment by Ghazgkull** [comment]
Did this issue make it into the 0.8 release? If not, is it too late to fix? This issue has continued to be a source of pain for us.

**Comment by evs-chris** [comment]
It's not in the 0.8 release, but I just had a weird idea - what if we used something like `document.currentScript` to check the url for query params on the script tag that is loading ractive, like `?RactiveGlobal=true`, and if that's there, set window.Ractive or global.Ractive or whatever. That way we could cover both sides of the global opinions without having to have special builds.
